Implement "Enemy unit sighted" notification

penev92 opened the pull request:
OpenRA currently has no such notification, and I think it needs it.
This adds a trait for the Player actor to track all neutral/enemy units that the player sees and a trait for units/structures/creeps to announce themselves when seen.
Currently added to D2k units and the sandworms.
Not added to D2k structures and TD/RA/TS mods because I don't have the appropriate notifications. If someone could assist me with those I'll add them.
Also removed the "WormSign" notification on worm spawn, since it never belonged there.

penev92 committed:
Implement EnemyWatcher and AnnounceOnSeen traits

Mailaender commented:
Needs a rebase.

Mailaender commented on mods/d2k/rules/player.yaml:
```diff
@@ -72,4 +72,4 @@ Player:
 	ProvidesTechPrerequisite@all:
 		Name: Unrestricted
 		Prerequisites: techlevel.low, techlevel.medium, techlevel.high, techlevel.superweapons
-
```
This newline removal will create unnecessary noise when we run our mod upgrade utility.

penev92 commented:
Rebased and updated with suggestions. Will deal with squashing at a later point.

Mailaender commented on mods/d2k/rules/player.yaml:
```diff
@@ -72,4 +72,5 @@ Player:
 		Prerequisites: techlevel.low, techlevel.medium, techlevel.high, techlevel.superweapons
-
+	EnemyWatcher:
+	
```
bogus indention

Mailaender commented:
I enabled developer cheats in d2k to trigger the worm sighting which did nothing. I disabled cheats again and tried to spot it manually, but then the game crashed:

```
Exception of type `System.NullReferenceException`: Object reference not set to an instance of an object
  at OpenRA.Mods.Common.Traits.EnemyWatcher.Tick (OpenRA.Actor self) [0x000ed] in OpenRA.Mods.Common/Traits/Player/EnemyWatcher.cs:74 
  at OpenRA.World.<Tick>m__3 (TraitPair`1 x) [0x0000e] in OpenRA.Game/World.cs:281 
  at (wrapper delegate-invoke) System.Action`1<OpenRA.TraitPair`1<OpenRA.Traits.ITick>>:invoke_void_T (OpenRA.TraitPair`1<OpenRA.Traits.ITick>)
  at OpenRA.WorldUtils.DoTimed[TraitPair`1] (IEnumerable`1 e, System.Action`1 a, System.String text) [0x00022] in OpenRA.Game/WorldUtils.cs:92 
  at OpenRA.World.Tick () [0x0013d] in OpenRA.Game/World.cs:281 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x001bf] in OpenRA.Game/Game.cs:458 
  at OpenRA.Game.LogicTick () [0x0004e] in OpenRA.Game/Game.cs:482 
  at OpenRA.Game.Loop () [0x000d4] in OpenRA.Game/Game.cs:594 
  at OpenRA.Game.Run () [0x00042] in OpenRA.Game/Game.cs:636 
  at OpenRA.Program.Run (System.String[] args) [0x00011] in OpenRA.Game/Support/Program.cs:111 
  at OpenRA.Program.Main (System.String[] args) [0x00050] in OpenRA.Game/Support/Program.cs:39
```

penev92 committed:
Add EnemyWatcher trait to D2k and AnnounceOnSeen to D2k units and sandworms; remove sandworm announcement on worm spawn

penev92 commented:
That didn't work with the developer cheat because of
`if (self.Owner.Shroud.Disabled)
	return;`
And the crash is because `actor.Actor.EffectiveOwner` is `null` (can someone explain the EffectiveOwner to me?).
Fixed indentation and crash, fixed the commits and rebased on top of #7278 (because this changes the `WormManager` trait).

penev92 commented:
Forgot to make worm sighting ping the radar, so I updated.

penev92 commented:
Rebased on top of bleed now.

Mailaender commented on OpenRA.Mods.Common/Traits/Player/EnemyWatcher.cs:
```diff
@@ -0,0 +1,117 @@
+			if (announce.Trait.Info.PingRadar && radarPings.Value != null)
+				radarPings.Value.Add(() => true, announce.Actor.CenterPosition, Color.Red, 50);
+
+			Game.Debug(announce.Trait.Info.Notification);
```
Do you want to keep this on purpose or is this left-over an accident?

penev92 replied:
Ah, I added this because I was testing on a PC where for some reason OpenRA's sound doesn't work, so I needed another way to keep track, other than breakpoints.
I think it would be better to remove it.

Mailaender commented:
:+1: Works as promised and is a nice refactor that makes the announcement easier to re-use.

penev92 commented:
Removed the debug messages.

obrakmann commented:
Looks good to me, :+1:

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/55593e1a43d96e09c49e832e2ab51e0d09d20dec%5E...55593e1a43d96e09c49e832e2ab51e0d09d20dec)
